### ISSUE TITLE
[source-postgres] JSONB field shows as string

### ISSUE THREAD

**Issue by adam-bloom** [airbytehq/airbyte#36327]
### Connector Name

source-postgres

### Connector Version

3.3.17

### What step the error happened?

Configuring a new connector

### Relevant information

#9028 is back!

1. Create a new connection using the latest postgresql source, from a source with a json/jsonb column in a table
2. Run schema discovery
3. See that column listed as a "string"

This causes major issues with some destinations (i.e. redshift) that have much smaller size limits for strings (65KB) than objects (16MB)

### Contribute

- [ ] Yes, I want to contribute

**Comment by Nazaniiin** [comment]
I am also seeing this behavior on Postgres connector 3.3.16

**Comment by marcosmarxm** [comment]
I think this is the current expected behavior. This is an extraction from the connector docs:
<img width="506" alt="image" src="https://github.com/airbytehq/airbyte/assets/5154322/1fc979b9-4171-4343-bb13-0ce94b08a9a5">

@airbytehq/dbsources do you think we should keep this open to further improvement to the connector?

**Comment by adam-bloom** [comment]
@marcosmarxm it looks like there was a change made a year ago that attempted to resolve this, but then it was perhaps removed? I linked to the original issue (#9028) where this was considered a bug, not an enhancement request. Even if this matches the docs, it results in limitations when using postgres for jsonb that don't exist with other sources. We moved some of our storage from mongodb to postgres, and the mongo connector can successfully sync rows that the postgres connector cannot due to its jsonb handling.

**Comment by Nazaniiin** [comment]
> @marcosmarxm it looks like there was a change made a year ago that attempted to resolve this, but then it was perhaps removed?

Yeah, it looks like it was reverted here https://github.com/airbytehq/airbyte/pull/23642 (there is a link to the comment with more context as to why it was reverted).

**Comment by marcosmarxm** [comment]
@airbytehq/dbsources not sure if this a change because of Java CDK or something. Can someone take a look in the next grooming section? The https://github.com/airbytehq/airbyte/pull/21695 made some changes to support JSONB data types. Thanks.

**Comment by evantahler** [comment]
We did some investigation on this ([internal doc](https://docs.google.com/document/d/1QL-7i2tKOSl7sh2P_ypXEl9tBPLM7vmmqSuBjggebKQ/edit)) and we believe that at the moment we can't support JSON/b types coming from relational database without adding a proper `json` datatype to the Airbyte Protocol. We do plan to add that late this year.

The problem is how we serialize data in transit with JSONschema. We would use `object` today, but both `'"foo"'` and `{"key": "value"}` are valid JSON according to postgres: 

```
✅ select '{"k": "v"}'::JSON ;
✅ select '[1,2,3]'::JSON ;
❌ select 'foo'::JSON ; // badly quoted
❌ select 1::JSONB;  // badly quoted
✅ select '"foo"'::JSON ;
✅ select '1'::JSONB ;
```

... and JSONSchema would reject the strings as invalid.  We'll be patching JSONSchema to have new custom JSON type which is the union of string and object.

**Comment by adam-bloom** [comment]
@evantahler is some interim support of full JSON only possible? I.e. opt in behavior instead of default? Or if this is required functionality do we need to build our own connector at this time?

**Comment by evantahler** [comment]
> @evantahler is some interim support of full JSON only possible? I.e. opt in behavior instead of default? Or if this is required functionality do we need to build our own connector at this time?

There isn't a way around this at this time.  Rather than build a new connector, can you cast the string column to JSON in a downstream step of your ELT pipeline, or build a view in the destination that does?

**Comment by adam-bloom** [comment]
> > @evantahler is some interim support of full JSON only possible? I.e. opt in behavior instead of default? Or if this is required functionality do we need to build our own connector at this time?
> 
> There isn't a way around this at this time. Rather than build a new connector, can you cast the string column to JSON in a downstream step of your ELT pipeline, or build a view in the destination that does?

@evantahler ideally, yes, but redshift has significant size limitations on string fields (65KB) that it does not have on objects. Other destinations, such as snowflake, support much larger strings, which would support a pattern that you're describing. In other words, our destination makes this a source constraint rather than something we can work around in the destination. The joys of Redshift! :)

**Comment by evantahler** [comment]
I _guess_ what could be done, if you wanted, is to make an advanced option for the connector one could toggle on which naively tries to cast postgres JSON types to Airybte/JSONSchema objects.  

Steps: 
* The default config should be disabled (be strings, like today).  Of, the choice would be of ENUMS (string and object)
* Add a warning that this is best effort, and will fail with any JSON record contents that aren't true objects (e.g strings, and probably even arrays)
* Change the code and give it a shot for your use case
* If it works, and we've made the option sufficiently scary, we could probably merge that in...

**Comment by adam-bloom** [comment]
@evantahler that sounds similar to my current postgres source, which is based off a 1.x version and naively casts JSON to objects. I think it's a reasonable approach.

Is that a change you're proposing that you all make, or that you want me to make and you're open to reviewing a PR?

**Comment by evantahler** [comment]
We aren't going to work on this in the near future, so if you want to add this option (with appropriate tests), we can merge it in

**Comment by makz81** [comment]
Are there any efforts to solve the problem here, or to add the option to cast the JSONB type, as described above? I am facing the same issue where very large JSON content in the Postgres source is getting truncated when transferred to Redshift. @adam-bloom , have you solved the problem in a different way?

**Comment by adam-bloom** [comment]
> Are there any efforts to solve the problem here, or to add the option to cast the JSONB type, as described above? I am facing the same issue where very large JSON content in the Postgres source is getting truncated when transferred to Redshift. @adam-bloom , have you solved the problem in a different way?

Hi @makz81 - I ended up solving the problem via a custom version of the postgres source connector that does the casting as described above. Unfortunately, it ended up being complex enough that it is not something an outside contributor like myself will easily be able to get clean enough to get into a PR, let alone merged. It required dipping far deeper than I was originally hoping into some of the airbyte internals to get the redshift destination type cast behavior to work correctly. @evantahler (and others on his team) were very helpful at providing some hints during that integration.

**Comment by octavia-squidington-iii** [comment]
At Airbyte, we seek to be clear about the project priorities and roadmap. This issue has not had any activity for 365 days, suggesting that it's not as critical as others. It's possible it has already been fixed. It is being marked as stale and will be closed in 20 days if there is no activity. To keep it open, please comment to let us know why it is important to you and if it is still reproducible on recent versions of Airbyte.

**Comment by adam-bloom** [comment]
This is definitely an important improvement to consider (at least given some destination limitations…ahem Redshift)

**Comment by bfayollevizeo** [comment]
This is key for us too. We have to use a nasty workaround I can't wait to ditch

**Comment by joerayme** [comment]
We've just encountered this issue as well and it might end up being a blocker for us adopting Airbyte (cloud). @bfayollevizeo curious to know what your workaround is, I can't think of anything that would work for us...

**Comment by bfayollevizeo** [comment]
> We've just encountered this issue as well and it might end up being a blocker for us adopting Airbyte (cloud). [@bfayollevizeo](https://github.com/bfayollevizeo) curious to know what your workaround is, I can't think of anything that would work for us...

We just restrict our json to less than 65535 characters based on some business logic. Far from ideal!

**Comment by hanslemm** [comment]
I fixed it in [my fork](https://github.com/hanslemm/airbyte) and published a [new image](https://github.com/hanslemm/airbyte/pkgs/container/airbyte%2Fsource-postgres) based on 3.7.0 Postgres Source with the fix, enough of limitations for Postgres sources JSON/JSONB.

I will later open a PR to the origin repo with a cleaner version of the solution.
